googledataprocauthenticator/magics/dataprocmagics.py

```diff
@@ -0,0 +1,238 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Runs Scala, PySpark and SQL statement through Spark using a REST endpoint in remote cluster.
+Provides the %spark and %manage_dataproc magics."""
+
+from IPython.core.magic import magics_class, line_cell_magic, needs_local_scope, line_magic
+from IPython.core.magic_arguments import argument, magic_arguments
+from hdijupyterutils.ipywidgetfactory import IpyWidgetFactory
+from sparkmagic.utils.utils import parse_argstring_or_throw, initialize_auth, Namespace
+from sparkmagic.livyclientlib.endpoint import Endpoint
+from sparkmagic.livyclientlib.exceptions import handle_expected_exceptions
+from sparkmagic.magics.remotesparkmagics import RemoteSparkMagics
+from sparkmagic.magics.sparkmagicsbase import SparkMagicBase
+from sparkmagic.controllerwidget.magicscontrollerwidget import MagicsControllerWidget
+import sparkmagic.utils.configuration as conf
+from sparkmagic.utils.constants import LANG_PYTHON, CONTEXT_NAME_SPARK, CONTEXT_NAME_SQL, LANG_SCALA, LANG_R
+
+
+@magics_class
+class DataprocMagics(SparkMagicBase):
+
+    def __init__(self, shell, data=None, widget=None):
+        super(DataprocMagics, self).__init__(shell, data)
+        self.ip = self.shell
+        self.db = self.ip.db
+        self.endpoints = {}
+        stored_endpoints = self.get_stored_endpoints()
+        for endpoint_tuple in stored_endpoints:
```

Comment:
Let's wrap this entire loop in a try-catch and log any failures to restore. (And also possibly consider clearing the stored endpoints if one occurs). If there's some failure case we hadn't anticipated, I don't want this to permanently fail the ability to even load the extension without manually clearing the DB.

Reply:
I added logging for exceptions in get_stored_endpoints and get_session_to_id https://github.com/GoogleCloudDataproc/dataprocmagic/pull/3/files/d0ee30418c5e1f2d47a5d377a1e43e5e9763a6dc#diff-28b0b312480d3bf4199e359119077b98R50. So originally I was only catching key error, which occurs because its the first time it is being stored, and just storing the variables as an empty list / dictionary. So I changed them to catch all errors and log the error and that it is  clearing them.